api/admin/controller/admin_auth_services.py

```diff
@@ -0,0 +1,88 @@
+from nose.tools import set_trace
+import flask
+from flask import Response
+from flask_babel import lazy_gettext as _
+from core.model import (
+    ExternalIntegration,
+    get_one,
+    get_one_or_create,
+)
+from core.util.problem_detail import ProblemDetail
+from . import SettingsController
+from api.admin.google_oauth_admin_authentication_provider import GoogleOAuthAdminAuthenticationProvider
+from api.admin.problem_details import *
+
+class AdminAuthServicesController(SettingsController):
+
+    def process_admin_auth_services(self):
+        self.require_system_admin()
+        provider_apis = [GoogleOAuthAdminAuthenticationProvider]
+        self.protocols = self._get_integration_protocols(provider_apis, protocol_name_attr="NAME")
```

Comment:
A method like this which is called on every request shouldn't modify instance variables like `.protocols`. The good news is that the protocols only need to be gathered once, and since you're making a separate class for each controller, you can set up the protocols for that controller in the constructor. So move this to a constructor and have it work something like this:

```
    def __init__(self, manager):
        super(AdminAuthServicesController, self).__init__(manager)
        provider_apis = [GoogleOAuthAdminAuthenticationProvider]
        self.protocols = self._get_integration_protocols(provider_apis, protocol_name_attr="NAME")
```

Reply:
Take a look at your existing classes and make sure they don't set `.protocols`. This is something I may have overlooked in previous reviews. If you do this going forward it'll all be good.